### PR TITLE
implement the EnvT data type and its instances

### PR DESCRIPTION
**Pull request by massimosiani**
I tried to implement the Env comonad, and its transformer EnvT, as the next to simplest example of a Comonad.
Thanks!

Thank you for contributing to Cats!

This is a kind reminder to run `sbt +prePR` and commit the changed files, if any, before submitting.

### CODE REVIEW

File: core/src/main/scala/cats/data/package.scala [hunk 1]
```diff
@@ -77,6 +77,11 @@ package object data extends ScalaVersionSpecificPackage {
   type RWS[E, L, S, A] = ReaderWriterState[E, L, S, A]
   val RWS = ReaderWriterState
 
+  type Env[E, A] = EnvT[Id, E, A]
+  object Env {
+    def apply[E, A](e: E, a: Id[A]): Env[E, A] = EnvT(a, e)
```
Comment (rossabaker) [inline core/src/main/scala/cats/data/package.scala]: Any reason not just `a: A` here?

The arguments are inconsistent with the order in `EnvT.apply`.

File: core/src/main/scala/cats/data/EnvT.scala [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+/**
+ * EnvT[W, E, A] is the Env Comonad transformer
+ */
+final case class EnvT[W[_], E, A](wa: W[A], env: E) {
+
+  /**
+   * Return the environment
+   */
+  lazy val ask: E = env
```
Comment (rossabaker) [inline core/src/main/scala/cats/data/EnvT.scala]: I don't think this needs to be lazy.  I would just make it a def.

File: core/src/main/scala/cats/data/EnvT.scala [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+  def local[E1](f: E => E1): EnvT[W, E1, A] = EnvT(wa, f(env))
+
+  def map[B](f: A => B)(implicit W: Functor[W]): EnvT[W, E, B] = EnvT(W.map(wa)(f), env)
+
+  lazy val runEnvT: (E, W[A]) = (env, wa)
+}
+
+object EnvT {
+
+  implicit def catsDataEnvTComonad[W[_]: Comonad, E]: Comonad[EnvT[W, E, *]] = new Comonad[EnvT[W, E, *]] {
```
Comment (rossabaker) [inline core/src/main/scala/cats/data/EnvT.scala]: You can get a `Functor[EnvT[W, E, *]]` from `Functor[W]`, if you want to split this up like `Foldable` and `Traversable`.

File: core/src/main/scala/cats/data/EnvT.scala [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+   * Transform the environment
+   */
+  def local[E1](f: E => E1): EnvT[W, E1, A] = EnvT(wa, f(env))
+
+  def map[B](f: A => B)(implicit W: Functor[W]): EnvT[W, E, B] = EnvT(W.map(wa)(f), env)
+
+  lazy val runEnvT: (E, W[A]) = (env, wa)
+}
+
+object EnvT {
```
Comment (rossabaker) [inline core/src/main/scala/cats/data/EnvT.scala]: There's an `Applicative[EnvT[W, E, *]]` from `Monoid[E]` and `Applicative[W]`.

File: core/src/main/scala/cats/data/EnvT.scala [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+  def asks[B](f: E => B): B = f(env)
+
+  /**
+   * Transform the environment
+   */
+  def local[E1](f: E => E1): EnvT[W, E1, A] = EnvT(wa, f(env))
+
+  def map[B](f: A => B)(implicit W: Functor[W]): EnvT[W, E, B] = EnvT(W.map(wa)(f), env)
+
+  lazy val runEnvT: (E, W[A]) = (env, wa)
```
Comment (rossabaker) [inline core/src/main/scala/cats/data/EnvT.scala]: I'd also skip the laziness here, though you do save a tuple if it's called repeatedly.

My gut feeling is these should be returned in the same order as the constructor.  I don't have a strong feeling which order is correct.

**Comment by massimosiani** [comment]
Hi @rossabaker and thanks for review!
Besides my interest in its own, comonads can be used for writing interpreters for (free) monads. For instance, I'd like to contribute by adding CofreeT, and the corresponding typeclasses in cats-mtl (ComonadStore and ComonadEnv for instance).

Sorry if was sloppy, probably the discussion doesn't fit in a PR comment :)

Let me know what you think about this!

File: core/src/main/scala/cats/data/package.scala (hunk 1 above)
Comment (massimosiani) [reply core/src/main/scala/cats/data/package.scala]: Well, the compiler complained, but a type annotation worked.

File: core/src/main/scala/cats/data/EnvT.scala (hunk 5 above)
Comment (massimosiani) [reply core/src/main/scala/cats/data/EnvT.scala]: Makes sense to me.

**Comment by larsrh** [comment]
The tests in this PR would need to be rewritten due to upstream changes. Please open a new PR if you're still interested in this feature.

**Pull request closed without merging**